src/main/java/io/pravega/perf/WriterWorker.java

```diff
@@ -57,67 +74,120 @@ public Void call() throws InterruptedException, ExecutionException, IOException
 
     private class EventsWriter implements Performance {
 
-        public void benchmark() throws InterruptedException, ExecutionException, IOException {
-            CompletableFuture retFuture = null;
-            Random rand = new Random();
+        public void benchmark() throws InterruptedException, IOException {
+            final EventsController eCnt = new EventsController(System.currentTimeMillis(), eventsPerSec);
+            for (int i = 0; i < events; i++) {
+                recordWrite(payload, stats::recordTime);
+                eCnt.control(i);
+            }
+            flush();
+        }
+    }
 
+    private class EventsWriterRW implements Performance {
+
+        public void benchmark() throws InterruptedException, IOException {
+            final EventsController eCnt = new EventsController(System.currentTimeMillis(), eventsPerSec);
             for (int i = 0; i < events; i++) {
+                final String val = System.currentTimeMillis() + ", " + workerID + ", ";
+                final String data = (val + payload).substring(0, messageSize);
```

Comment:
You can get the `payload` of the appropriate size outside the loop, as the format of the String `val` is constant. This will make each iteration more efficient, as you could avoid executing `substring()`, just appending the `val` to the to the `payload`. Moreover, you could also add an assertion to verify that the payload size is the expected one.

Reply:
yes, i have done the payload contruction outside the loop for methods EventsWriter , EventsWriterTime  but you are seeing the methods EventsWriterRW and EventsWriterTimeRW which are required for end to end latency.  for end to end latency, the start time will be existing in every event sent and so that reader reads its and marks it as a start time.

Reply:
My point is not related to the fact that you need to calculate the timestamp for each message sent to get the end-to-end latency. Instead, I wonder about if we could make more efficient this line:
`final String data = (val + payload).substring(0, messageSize);`
This is doing a string concatenation and a call to `substring()` method per iteration, which is expensive. I suggest to explore the use `StringBuffer` for concatenation and avoid to operate as much as possible on `payload` within that loop as it may be up to 1MB in size.

Reply:
now, I have used the StringBuilder to construct the payload for end to end latency. only the hearder is changed for every write.